fivox/cuda/simpleLFP.cu

```diff
@@ -0,0 +1,128 @@
+
+/* Copyright (c) 2014-2016, EPFL/Blue Brain Project
+ *                          Jafet.VillafrancaDiaz@epfl.ch
+ *
+ * This file is part of Fivox <https://github.com/BlueBrain/Fivox>
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License version 3.0 as published
+ * by the Free Software Foundation.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include "simpleLFP.h"
+
+namespace cuda
+{
+__global__ void kernelLFP( const float* __restrict__ eventsX,
+                           const float* __restrict__ eventsY,
+                           const float* __restrict__ eventsZ,
+                           const float* __restrict__ radii,
+                           const float* __restrict__ values,
+                           const Parameters& params, const VolumeInfo& volInfo,
+                           float* __restrict__ output )
+{
+    // 1D grid of 1D blocks
+    const int threadId = blockIdx.x * blockDim.x + threadIdx.x;
+
+    const int xIndex = threadId % volInfo.dimensions.x;
+    const int yIndex = ( threadId / volInfo.dimensions.x )
+                       % volInfo.dimensions.y;
+    const int zIndex = (( threadId / volInfo.dimensions.x )
+                        / volInfo.dimensions.y ) % volInfo.dimensions.z;
+
+    const float voxelPosX = xIndex * volInfo.voxelSize + volInfo.origin.x;
+    const float voxelPosY = yIndex * volInfo.voxelSize + volInfo.origin.y;
+    const float voxelPosZ = zIndex * volInfo.voxelSize + volInfo.origin.z;
+
+    // Compute directly the inverted value to gain performance in the for loop
+    const float cutOffDistance = 1.f / params.cutoff;
+
+    float current( 0.f );
+    for( unsigned int i = 0; i < params.numEvents; ++i )
+    {
+        const float value( values[i] );
```

Comment:
Here there's an optimization opportunity. Instead of reading values[i] from each thread, you could read a a different value to shared memory (and the same for the event xyz coordinates). Even more interesting would be to compare the performance between using shared memory and warp shuffle operations.